### PR TITLE
8314502: Change the comparator taking version of GrowableArray::find to be a template method

### PR DESCRIPTION
**Pull request by afshin-zafari**
The `find` method now is 
```C++
template<typename T>
int find(T* token, bool f(T*, E)) const {
...
```
Any other functions which use this are also changed.
Local linux-x64-debug hotspot:tier1 passed. Mach5 tier1 build on linux and Windows passed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314502](https://bugs.openjdk.org/browse/JDK-8314502): Change the comparator taking version of GrowableArray::find to be a template method (**Enhancement** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [266f6feb](https://git.openjdk.org/jdk/pull/15418/files/266f6feb4a2e045a71bf447c0a9f0fd497e252c5)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [6d9288e7](https://git.openjdk.org/jdk/pull/15418/files/6d9288e7e9c005dd98013c1e6a85450da61fbd88)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15418/head:pull/15418` \
`$ git checkout pull/15418`

Update a local copy of the PR: \
`$ git checkout pull/15418` \
`$ git pull https://git.openjdk.org/jdk.git pull/15418/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15418`

View PR using the GUI difftool: \
`$ git pr show -t 15418`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15418.diff">https://git.openjdk.org/jdk/pull/15418.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15418#issuecomment-1691772262)

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiImpl.cpp
Comment (jdksjolen) [inline src/hotspot/share/prims/jvmtiImpl.cpp:126]: Please rename the `v` parameter to `e1`

Comment (jdksjolen) [reply src/hotspot/share/prims/jvmtiImpl.cpp:126]: And since you're here: Could you change the code style so that it's `GrowableElement* e2` and not `GrowableElement *e2`?

File: src/hotspot/share/prims/jvmtiImpl.hpp
Comment (jdksjolen) [inline src/hotspot/share/prims/jvmtiImpl.hpp:91]: Remove spacing between `GrowableElement` and `*`

**Review by jdksjolen — APPROVED** [review]
Looks good to me, thank you. A couple of style issues that needs to be fixed, some pre-existing.

File: src/hotspot/share/utilities/growableArray.hpp
Comment (dholmes-ora) [inline src/hotspot/share/utilities/growableArray.hpp:213]: Pardon my ignorance here, but what is the type relationship between T and E?

File: src/hotspot/share/prims/jvmtiImpl.cpp
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiImpl.cpp:126]: Fixed.

File: src/hotspot/share/prims/jvmtiImpl.hpp
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiImpl.hpp:91]: Fixed.

File: src/hotspot/share/utilities/growableArray.hpp
Comment (jdksjolen) [reply src/hotspot/share/utilities/growableArray.hpp:213]: It's arbitrary and chosen by the caller through `f`, so I can't say :-). The best use case we have now is when you only have an `int` which uniquely describes an `LGRPSpace`.

Comment (stefank) [reply src/hotspot/share/utilities/growableArray.hpp:213]: This code looks similar to a capturing lambda. Would an alternative be to use that instead and let the calling code be changed to:

```
  int i = lgrp_spaces()->find([](LGRPSpace* space) {
    return space->lgrp_id() == lgrp_id;
  });
```

Alternatively:
```
  auto matches_lgrp_id = [](LGRPSpace* space) {
    return space->lgrp_id() == lgrp_id;
  };
  
  int i = lgrp_spaces()->find(matches_lgrp_id);
```

Comment (jdksjolen) [reply src/hotspot/share/utilities/growableArray.hpp:213]: We could just as well do a capturing lambda here, yes. Then we'd have:

```c++
template<F>
int find(F finder);
```

It'd be a template instead of function pointer since it's a capturing lambda and  `std::function` is not permitted in Hotspot AFAIK. 

As an aside, to clarify for readers: There's a `&` missing in the capture list of your examples.

Comment (stefank) [reply src/hotspot/share/utilities/growableArray.hpp:213]: Updated in-place

Comment (kimbarrett) [reply src/hotspot/share/utilities/growableArray.hpp:213]: Not a review, just agreeing with @stefank and @jdksjolen .  What they describe is idiomatic C++.

Comment (kimbarrett) [reply src/hotspot/share/utilities/growableArray.hpp:213]: Also, why isn't this change also being applied to `find_from_end`

**Comment by merykitty** [comment]
This is similar to `std::find_if` and should be just:

      template<class UnaryPredicate>
      int find(UnaryPredicate p) const {
        for (int i = 0; i < _len; i++) {
          if (p(_data[i])) {
            return i;
          }
        }
        return -1;
      }

Regarding the current approach, the comparator should take a `const E&` instead of an `E`, and the token passed in should be `const` also.

File: src/hotspot/share/utilities/growableArray.hpp
Comment (sspitsyn) [reply src/hotspot/share/utilities/growableArray.hpp]: There can be a confusion related to selection of type names T and E:
  T is intuitively treated as a table and E as an element.
No pressure but I wonder if using D instead of T would be better.

**Comment by afshin-zafari** [comment]
> This is similar to `std::find_if` and should be just:
> 
> ```
>   template<class UnaryPredicate>
>   int find_if(UnaryPredicate p) const {
>     for (int i = 0; i < _len; i++) {
>       if (p(_data[i])) {
>         return i;
>       }
>     }
>     return -1;
>   }
> ```
> 
> Regarding the current approach, the comparator should take a `const E&` instead of an `E`, and the token passed in should be `const` also.

Thank you  @merykitty for this comment. 
It is fixed now.

File: src/hotspot/share/utilities/growableArray.hpp
Comment (afshin-zafari) [reply src/hotspot/share/utilities/growableArray.hpp]: > Also, why isn't this change also being applied to `find_from_end`

Thank you @kimbarrett, the function is also changed accordingly.

Comment (afshin-zafari) [reply src/hotspot/share/utilities/growableArray.hpp]: > We could just as well do a capturing lambda here, yes. Then we'd have:
> 
> ```c++
> template<F>
> int find(F finder);
> ```
> 
> It'd be a template instead of function pointer since it's a capturing lambda and `std::function` is not permitted in Hotspot AFAIK.
> 
> As an aside, to clarify for readers: There's a `&` missing in the capture list of your examples.

It would be nice to have this templated Function as finder. However, I think it is better to keep the changes small and manageable for this PR. Thanks for the comment.

Comment (afshin-zafari) [reply src/hotspot/share/utilities/growableArray.hpp]: > Also, why isn't this change also being applied to `find_from_end`

Thanks Kim! It is also changed.

**Comment by afshin-zafari** [comment]
Dear reviewers @kimbarrett, @dholmes-ora, @stefank, @sspitsyn and @merykitty, 
Would you please check if you have any more comments on this? Thanks.

File: src/hotspot/share/gc/parallel/mutableNUMASpace.hpp
Comment (dholmes-ora) [inline src/hotspot/share/gc/parallel/mutableNUMASpace.hpp:1]: This seems an unrelated change.

Comment (afshin-zafari) [reply src/hotspot/share/gc/parallel/mutableNUMASpace.hpp:1]: This change came after fixing a merge conflict.
In `mutableNUMASpace.cpp`, at lines 163, 182, 202 and 586 the `find` function is called in this way: 
```cpp
int i = lgrp_spaces()->find(&lgrp_id, LGRPSpace::equals);
```
where `lgrp_id` is `int`. Therefore, the `LGRPSpace::equals` has to take an `int*` in its first argument. The definition of `find` is:
```cpp
int find(T* token, bool f(T*, const E&)) const {
```

Comment (dholmes-ora) [reply src/hotspot/share/gc/parallel/mutableNUMASpace.hpp:1]: After JDK-8316115 `lgrp_id` is `uint`.

**Comment by sspitsyn** [comment]
The serviceability files look good.
By being paranoid I'd suggest to run more tiers, eg. 3-4.

**Review by sspitsyn — APPROVED** [review]

**Comment by afshin-zafari** [comment]
> The serviceability files look good. By being paranoid I'd suggest to run more tiers, eg. 3-4.

tiers 1-4 are passed.
Thanks for your advice.

**Comment by afshin-zafari** [comment]
@dholmes-ora, any more comments?

**Comment by dholmes-ora** [comment]
@afshin-zafari  could you merge (not rebase!) with current master please to ensure I can see the absolute latest version of this patch in relation to master. Thanks.

**Comment by afshin-zafari** [comment]
> @afshin-zafari could you merge (not rebase!) with current master please to ensure I can see the absolute latest version of this patch in relation to master. Thanks.

It is up to date with master and nothing to push.

**Comment by dholmes-ora** [comment]
> It is up to date with master and nothing to push.

That doesn't seem possible since you last merged with master on September 26.

**Comment by afshin-zafari** [comment]
> > It is up to date with master and nothing to push.
> 
> That doesn't seem possible since you last merged with master on September 26.

Now it is up to date after sync'ing my fork.

File: src/hotspot/share/utilities/growableArray.hpp
Comment (dholmes-ora) [inline src/hotspot/share/utilities/growableArray.hpp:213]: What is the advantage of a const reference here?

Comment (merykitty) [reply src/hotspot/share/utilities/growableArray.hpp:213]: You can bind a non-const reference to a const one but not the other way.

Comment (dholmes-ora) [reply src/hotspot/share/utilities/growableArray.hpp:213]: Sorry I was unclear: what is the advantage of a reference here? Is it just to avoid copying

Comment (merykitty) [reply src/hotspot/share/utilities/growableArray.hpp:213]: @dholmes-ora Yes it helps avoid copying, especially if the copy constructor is non-trivial. And I think it is more idiomatic in C++ to use references here.

Comment (kimbarrett) [reply src/hotspot/share/utilities/growableArray.hpp:213]: Using a reference here leads to unnecessary overhead when `E` is small and
trivially copyable, unless the predicate function is inlined. Pass by value is
better in that case. Of course, as noted above, if `E` is "expensive" to copy
or non-copyable then a reference is needed. Boost has this thing called
`call_traits<E>::param_type` for this issue, but I don't recommend we copy
that.

Idiomatic C++ makes the entire function a template parameter, as was suggested
earlier in this PR. That dodges this question entirely, leaving the parameter
passing decision to the predicate function where it belongs, rather than
having it imposed by GrowableArray::find.  The find function just imposes the
requirement that the predicate satisfies the appropriate constraints, e.g. it
is callable on an element reference and returns convertible to bool.

**Comment by kimbarrett** [comment]
> I still approve of this patch as it's better than what we had before. There are a lot of suggested improvements that can be done either in this PR or in a future RFE. `git blame` shows that this hasn't been touched since 2008, so I don't think applying all suggestions now is in any sense critical :-).

Not touched since 2008 suggests to me there might not be a rush to make the change as proposed, and instead take
the (I think small) additional time to do the better thing, e.g. the unary-predicate suggestion made by several folks.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
I would prefer the function template with a unary predicate approach suggested earlier
by multiple commenters. Is there a reason not to just do that?

File: src/hotspot/share/utilities/growableArray.hpp
Comment (rose00) [reply src/hotspot/share/utilities/growableArray.hpp:213]: I agree we should be using a template-typed function instead of a function pointer here.
I think a lot of our uses of function pointers in our code base would work better as template-typed args.
See for example the `grow` argument (of template type `GFN`) at this point:
https://github.com/openjdk/jdk/blob/d051f22284e7ccc288c658588f73da672d9bfacd/src/hotspot/share/utilities/unsigned5.hpp#L343C34-L343C34

In that case I cited, if the `grow` argument it were a function pointer instead of a template-typed function-like argument (Kim what’s the right term here? “functor”??), then performance and flexibility would be unacceptably harmed.

I think the `find` argument is just the same kind of thingy.

Comment (kimbarrett) [reply src/hotspot/share/utilities/growableArray.hpp:213]: "Functor" is a reasonable choice, with plenty of precedent. 

The C++ standard doesn't use the term "functor". Instead it defines (C++14
20.9) the terms "function object type" and associated "function object" to
include all of

(1) pointer to function types
(2) class types with a member operator()
(3) class types with a conversion to pointer to function

The C++ standard doesn't have distinct terminology for cases (2) and (3), so
far as I can tell.

Sometimes people use either "functor" or "function object class" to
distinguish (2) (web search for "C++ functor" for examples of the former).
Sometimes people use "function object" to refer only to (2). So the community
is inconsistent in this respect.

I'm not sure I've ever seen (3).  (Not even in "look at this obscure C++
thing" discussions.)

**Comment by afshin-zafari** [comment]
@kimbarrett , @dholmes-ora , @merykitty 
Is there any comment on this PR?

File: src/hotspot/share/gc/parallel/mutableNUMASpace.cpp
Comment (dholmes-ora) [inline src/hotspot/share/gc/parallel/mutableNUMASpace.cpp:163]: I guess it is somewhat outside the scope of this PR but I wish this code would make its mind up about whether the NUMA group ids are `int` or `uint`!

**Comment by dholmes-ora** [comment]
@afshin-zafari I will leave it to other to (re-) review the latest changes. I don't grok this template stuff enough to pass judgement.

File: src/hotspot/share/utilities/growableArray.hpp
Comment (merykitty) [inline src/hotspot/share/utilities/growableArray.hpp:213]: Should be

    template <typename F>
    int find(F f) const {
        for (int i = 0; i < _len; i++) {
            if (f(_data[i]) {
                return i;
            }
        }
        return -1;
    }

Comment (afshin-zafari) [reply src/hotspot/share/utilities/growableArray.hpp:213]: We need `token` to find it in the array, don't we? All the invocations pass such a function with two parameters. The change here needs all invocations to be changed.

**Comment by afshin-zafari** [comment]
> @afshin-zafari I will leave it to other to (re-) review the latest changes. I don't grok this template stuff enough to pass judgement.

Thank you very much @dholmes-ora, for your comments and discussions.

**Comment by stefank** [comment]
@afshin-zafari I think you misunderstand the feedback given. The suggestions from many of us has been that you should change the functions to accept lambdas / template-typed function instead.

https://github.com/openjdk/jdk/pull/15418#discussion_r1305389552
https://github.com/openjdk/jdk/pull/15418#discussion_r1375386257
https://github.com/openjdk/jdk/pull/15418#discussion_r1376940244
https://github.com/openjdk/jdk/pull/15418#discussion_r1395219935

File: src/hotspot/share/utilities/growableArray.hpp
Comment (merykitty) [reply src/hotspot/share/utilities/growableArray.hpp:213]: No, it can be embedded into the function object.

Comment (merykitty) [reply src/hotspot/share/utilities/growableArray.hpp:213]: I think you can have 2 versions, `GrowableArray::find(const E& value)` and `GrowableArray::find_if(UnaryPredicate p)`.

**Comment by stefank** [comment]
Thanks for making this change.

I'd like to suggest the following cleanups, some documentation, and a few tests:
https://github.com/openjdk/jdk/commit/20d4502471ba396ae395512cfa3dab3f87555421

I think it might be easier to review by looking at the final diff:
https://github.com/openjdk/jdk/compare/master...stefank:jdk:pr_15418

**Comment by afshin-zafari** [comment]
> Thanks for making this change.
> 
> I'd like to suggest the following cleanups, some documentation, and a few tests: [20d4502](https://github.com/openjdk/jdk/commit/20d4502471ba396ae395512cfa3dab3f87555421)
> 
> I think it might be easier to review by looking at the final diff: [master...stefank:jdk:pr_15418](https://github.com/openjdk/jdk/compare/master...stefank:jdk:pr_15418)

One question: the `private static bool by_name(const char* name, PerfData* pd);` is added to `PerfDataList` class but is never used.  Is something missing?

**Comment by stefank** [comment]
> > Thanks for making this change.
> > I'd like to suggest the following cleanups, some documentation, and a few tests: [20d4502](https://github.com/openjdk/jdk/commit/20d4502471ba396ae395512cfa3dab3f87555421)
> > I think it might be easier to review by looking at the final diff: [master...stefank:jdk:pr_15418](https://github.com/openjdk/jdk/compare/master...stefank:jdk:pr_15418)
> 
> One question: the `private static bool by_name(const char* name, PerfData* pd);` is added to `PerfDataList` class but is never used. Is something missing?

That should have been removed when I added name_equals.

**Comment by afshin-zafari** [comment]
@stefank,  the changes are applied as suggested and ready for review.

**Comment by stefank** [comment]
@afshin-zafari you toke my changes and then you made a few small whitespace changes that messed up the patch. I merged our two branches to correct those issues. Could you make sure to pull the changes from my branch (don't copy the changes):
https://github.com/stefank/jdk/tree/pr_15418

**Review by stefank — APPROVED** [review]

**Comment by afshin-zafari** [comment]
Dear @kimbarrett , @dholmes-ora , @stefank, @rose00 , @jdksjolen, @sspitsyn, @merykitty, 
Thank you all. 
The final code is now cleaner and nicer.
/integrate

**Comment by dholmes-ora** [comment]
@afshin-zafari I'm happy to hear the code is now cleaner and nicer, but AFAICS this new version of the code has only had a single review. It should really have been re-reviewed by at least one of the earlier reviewers. Thanks.

**Comment by afshin-zafari** [comment]
> @afshin-zafari I'm happy to hear the code is now cleaner and nicer, but AFAICS this new version of the code has only had a single review. It should really have been re-reviewed by at least one of the earlier reviewers. Thanks.

Oh, really sorry. That's obviously my mistake. Your are right, I should have waited for others.
What to do now?

**Comment by stefank** [comment]
> > @afshin-zafari I'm happy to hear the code is now cleaner and nicer, but AFAICS this new version of the code has only had a single review. It should really have been re-reviewed by at least one of the earlier reviewers. Thanks.
> 
> Oh, really sorry. That's obviously my mistake. Your are right, I should have waited for others. What to do now?

I think the other reviewers can give their review feedback then you create a new RFE to fix whatever they'd like to get fixed.

**Comment by jdksjolen** [comment]
Still LGTM.

**Pull request closed without merging**
